Use IsDisturbed(ReadableStream)

yutakahirano opened the pull request:
This fixes #45. The link should be modified after IsDisturbed is specified on the streams spec.

annevk commented:
I see, the "passed flag" is a renaming of the "used flag". Perhaps we should call it "obsolete flag"? I don't think "passed" is very clear.

Also, we need to change `Request`'s used predicate, since it needs to check for IsDisturbed.

I also think the "locked flag" can go, since we were going to define an abstract operation in Streams for `text()` etc. that would mark the stream disturbed immediately.

yutakahirano commented:
@annevk Thanks, I uploaded a revised one.

> I see, the "passed flag" is a renaming of the "used flag". Perhaps we should call it "obsolete flag"? I don't think "passed" is very clear.

It came from https://github.com/slightlyoff/ServiceWorker/issues/452 and yes that should be renamed. I think it's ok to call it as "used flag".

> Also, we need to change Request's used predicate, since it needs to check for IsDisturbed.

We will, but currently Request doesn't have its body stream. `used flag` emulates the effect. Hence I'm using the flag for now.

> I also think the "locked flag" can go, since we were going to define an abstract operation in Streams for text() etc. that would mark the stream disturbed immediately.

Fixed.

annevk commented on README.md:
```diff
@@ -118,22 +118,18 @@ The following item is deleted.
-1. If passed flag is set or locked flag is set, return a new promise rejected with a TypeError.
-1. Set locked flag.
+1. If this Request is used, return a new promise rejected with a TypeError.
+1. If request body is not empty, set used flag.
```
Is not empty -> is non-null, right?

yutakahirano replied:
non-null should be checked. Done.
I think an already-closed empty stream cannot be disturbed and hence used flag should not be set here, right?

annevk replied:
Actually, looking at this in more context this looks wrong. Consume body should always set the used flag. As it will always result in a disturbed stream. An already-closed empty stream would already be disturbed I think. Though @domenic can say for sure.

yutakahirano commented:
I changed my mind: I decoupled disturbed flag from used flag. The former emulates IsDisturbed(stream) and the latter is defined in https://etherpad.mozilla.org/streams-f2f-july:L45 and L46.

annevk commented on README.md:
```diff
@@ -118,22 +118,20 @@ The following item is deleted.
-
-- passed flag is set.
-- locked flag is set.
+A Request has an associated __used__ predicate that returns true if one of disturbed flag and used flag is set.
```
Use "if either the disturbed flag or the used flag is set". "One of" can be used when you have more than two.

yutakahirano replied:
Done.

annevk commented on README.md:
```diff
@@ -118,22 +118,20 @@ The following item is deleted.
-1. If passed flag is set or locked flag is set, return a new promise rejected with a TypeError.
-1. Set locked flag.
+1. If this Request is used, return a new promise rejected with a TypeError.
+1. If request body is non-null and not empty, set disturbed flag.
```
I don't think the empty check makes sense. We can't tell whether a stream is empty. The conclusion we reached was that Streams would define a "consume stream" algorithm of sorts that would immediately set the disturbed flag.

yutakahirano replied:
Does that mean you changed the policy after http://krijnhoetmer.nl/irc-logs/whatwg/20150722#l-459? If so, I'll fix the change.

annevk replied:
The idea is that a stream becomes disturbed if `read()` or `cancel()` is invoked on it. This will always happen for "consume body", unless body is null. When body is null, it will always succeed (due to treating that as an explicit empty byte sequence).

@wanderview and @domenic can confirm this (I hope!).

yutakahirano replied:
Done.

yutakahirano commented:
Addressed review comments. My current understanding is:

- There is no clear distinction between an evident empty stream (i.e. an already closed empty stream) and a maybe empty stream. Both will be disturbed when read() / cancel() is called.
- There is a clear distinction between a null stream and an empty stream. The former cannot be disturbed while the latter can be.

Please correct me if I'm wrong.

annevk commented:
That matches my understanding.

domenic commented:
Mine as well.
